openedx/core/djangoapps/course_roles/data.py

```diff
@@ -0,0 +1,160 @@
+"""
+Permissions for course roles app.
+
+These are the permissions that can be assigned to a CourseRole which grants access for a course,
+but can also be assigned for org wide course access or instance wide course access.
+They are defined in the database in the course_roles_permission table.
+
+To add a new permission, add a new entry to the CourseRolesPermission enum,
+then and add a new row to the course_roles_permission table in database,
+with a migration.
+
+To remove a permission, remove the entry from the CourseRolesPermission enum,
+and remove the row from the course_roles_permission table in database,
+with a migration.
+
+To change the readable_name or description of a permission, change the
+corresponding entry in CourseRolesPermission enum.
+
+To change the name of a permission, change the corresponding entry on the
+CourseRolesPermission enum, and change the name field of the corresponding
+row in the course_roles_permission table in database, with a migration.
+"""
+from attrs import frozen, field, validators
+from enum import Enum, unique
+
+from django.utils.translation import gettext as _
+
+
+@frozen
+class PermissionData:
+    """
+    Data class for a permission.
+    """
+    name: str = field(validator=validators.instance_of(str))
+    readable_name: str = field(validator=validators.instance_of(str))
+    description: str = field(validator=validators.instance_of(str))
+
+
+@unique
+class CourseRolesPermission(Enum):
+    """
+    Enum of all user permissions, the values are the permissions names
+    in the course_roles_permission table in database.
+
+    The readable_name and description are used in the UI.
+    """
+
+    MANAGE_CONTENT = PermissionData(
+        "manage_content",
+        _("Manage Content"),
+        _("Can view, create, edit, delete and publish (not publisher tool) any course content."),
+    )
+    MANAGE_COURSE_SETTINGS = PermissionData(
+        "manage_course_settings", _("Manage Course Settings"), _("Can view and edit settings pages.")
+    )
+    MANAGE_ADVANCED_SETTINGS = PermissionData(
+        "manage_advanced_settings", _("Manage Advanced Settings"), _("Can view and edit advanced settings.")
+    )
+    VIEW_COURSE_SETTINGS = PermissionData(
+        "view_course_settings", _("View Course Settings"), _("Can view all settings pages.")
+    )
+    VIEW_ALL_CONTENT = PermissionData(
+        "view_all_content",
+        _("View All Content"),
+        _(
+            "Can view course content in LMS in all statuses: "
+            "Published and live, Draft, Staff-only, Published-not-yet-released."
+        ),
+    )
+    VIEW_ONLY_LIVE_PUBLISHED_CONTENT = PermissionData(
+        "view_only_live_published_content",
+        _("View Only Live Published Content"),
+        _(
+            "Can only view published-and-live content in a course. "
+            "Cannot view published-not-yet-released content. Cannot view Staff-only content."
```

Comment:
Nit: It seems odd to have a permission that tells you what you can't do. It looks like this is supposed to be exclusive with `VIEW_ALL_CONTENT` so that you only have one or the other, but I would have expected this permission to be `VIEW_LIVE_PUBLISHED_CONTENT` and for staff to have both.

Reply:
A role can have both permissions, but in that case `VIEW_ONLY_LIVE_PUBLISHED_CONTENT` will not actually add any access. All content includes all live published content so having both would be a redundancy, but it won't cause any issues in the code. 

This permission (`VIEW_ONLY_LIVE_PUBLISHED_CONTENT`) exists so that it can be given to users who should only see live published content and not published content before it is live or unpublished content. 

@cablaa77 Please let us know if you agree the permission name should be changed to `VIEW_LIVE_PUBLISHED_CONTENT` from `VIEW_ONLY_LIVE_PUBLISHED_CONTENT`.

Reply:
I guess the general pattern I am leery of is that people will start to use a permission check as a proxy for the absence of other permissions, and that it will introduce unnecessary coupling when we want to change things later. But I don't have a concrete case off the top of my head, and I won't block approving this PR over this.

Reply:
The permission name is being changed to `VIEW_LIVE_PUBLISHED_CONTENT`